src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,834 @@
+                        preZi[21] = (byte) i;
+                        lmOtSig.getY(i, preZi, 23);
+                    }
+
+                    for (int j = a; j < twoPowWMinus1; j++) {
+                        preZi[22] = (byte) j;
+                        if (j < twoPowWMinus2) {
+                            digestFixedLengthPreprocessed(
+                                    sha256, preZi, hashLen, preZi, 23, n);
+                        } else {
+                            digestFixedLengthPreprocessed(
+                                    sha256, preZi, hashLen, preCandidate, 22 + i * n, n);
+                        }
+                    }
+                }
+                md.update(preCandidate, 0, 22 + p * n);
+
+                result = md.digest();
+            } catch (DigestException e) { // This should not happen
+                throw new ProviderException("Digest failed", e);
+            }
+            if (n != hashAlg_n) {
+                result = Arrays.copyOfRange(result, 0, n);
+            }
+            return result;
+        }
+    }
+
+    public static class KeyFactoryImpl extends KeyFactorySpi {
+        @Override
+        protected PublicKey engineGeneratePublic(KeySpec keySpec)
+                throws InvalidKeySpecException {
+            if (keySpec instanceof X509EncodedKeySpec) {
+                try {
+                    X509EncodedKeySpec x509Spec = (X509EncodedKeySpec)keySpec;
+                    return new HSSPublicKey(
+                            x509Spec.getEncoded(), true);
+                } catch (InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            } else if (keySpec instanceof RawKeySpec x) {
+                try {
+                    return new HSSPublicKey(x.getKeyArr(), false);
+                } catch (InvalidKeyException e) {
+                    throw new InvalidKeySpecException(e);
+                }
+            }
+            throw new InvalidKeySpecException("Unrecognized KeySpec");
+        }
+
+        @Override
+        protected PrivateKey engineGeneratePrivate(KeySpec keySpec)
+                throws InvalidKeySpecException {
+            throw new InvalidKeySpecException(
+                    "Private key generation is not supported");
+        }
+
+        @Override
+        protected <T extends KeySpec> T engineGetKeySpec(Key key, Class<T> keySpec)
+                throws InvalidKeySpecException {
+            if (key == null) {
+                throw new InvalidKeySpecException("key should not be null");
+            }
+            if (key.getFormat().equals("X.509") &&
+                    key.getAlgorithm().equalsIgnoreCase("HSS/LMS")) {
+                if (keySpec.isAssignableFrom(X509EncodedKeySpec.class)) {
+                    return keySpec.cast(new X509EncodedKeySpec(key.getEncoded()));
+                }
+                throw new InvalidKeySpecException("keySpec is not an X509 one");
```

Comment:
Suggest saying name of class: "keySpec is not an X509EncodedKeySpec"

Reply:
Said.